storage/azblob/azblob.go

```diff
@@ -0,0 +1,391 @@
+// Copyright 2022 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package azblob
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	"github.com/mendersoftware/deployments/model"
+	"github.com/mendersoftware/deployments/storage"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+)
+
+const (
+	headerBlobType             = "x-ms-blob-type"
+	headerMSContentDisposition = "x-ms-blob-content-disposition"
+
+	blobTypeBlock = "BlockBlob"
+)
+
+type client struct {
+	*azblob.ContainerClient
+	fileSuffix  *string
+	contentType *string
+	bufferSize  int
+}
+
+type OpError struct {
+	Op      string
+	Message string
+	Reason  error
+}
+
+func (err OpError) Error() string {
+	errStr := "azblob"
+	if err.Op != "" {
+		errStr += " " + err.Op
+	}
+	if err.Message != "" {
+		errStr += ": " + err.Message
+	}
+	if err.Reason != nil {
+		errStr += ": " + err.Reason.Error()
+	}
+	return errStr
+}
+
+func (err OpError) Unwrap() error {
+	return err.Reason
+}
+
+type SharedKeyCredentials struct {
+	AccountName string
+	AccountKey  string
+
+	URI *string // Optional
+}
+
+func New(ctx context.Context, bucket string, opts ...*Options) (storage.ObjectStorage, error) {
+	var (
+		err error
+		cc  *azblob.ContainerClient
+	)
+	opt := NewOptions(opts...)
+	if opt.ConnectionString != nil {
+		cc, err = azblob.NewContainerClientFromConnectionString(
+			*opt.ConnectionString, bucket, &azblob.ClientOptions{},
+		)
+		if err != nil {
+			return nil, err
+		}
+	} else if sk := opt.SharedKey; sk != nil {
+		var cred *azblob.SharedKeyCredential
+		cred, err = azblob.NewSharedKeyCredential(sk.AccountName, sk.AccountKey)
+		if err != nil {
+			return nil, err
+		}
+		var containerURI string
+		if sk.URI != nil {
+			containerURI = *sk.URI
+		} else {
+			containerURI = fmt.Sprintf(
+				"https://%s.blob.core.windows.net/%s",
+				cred.AccountName(),
+				bucket,
+			)
+		}
+		cc, err = azblob.NewContainerClientWithSharedKey(
+			containerURI,
+			cred,
+			&azblob.ClientOptions{},
+		)
+	}
+	if err != nil {
+		return nil, err
+	}
+	objectStorage := &client{
+		ContainerClient: cc,
+		fileSuffix:      opt.FilenameSuffix,
+		contentType:     opt.ContentType,
+	}
+	if err := objectStorage.HealthCheck(ctx); err != nil {
```

Comment:
is the HealthCheck also applicable to the service health check?

Reply:
Yes, it is already included in the liveliness api endpoint.

Reply:
This is just a startup check to verify the credentials have access to the container/bucket (and that it exist).